### DISCUSSION TITLE
Questions about `framing_sv2` crate

### DISCUSSION THREAD

**Discussion by plebhash** [stratum-mining/stratum#858] [General]
I'm working on the documentation effort described on https://github.com/stratum-mining/stratum/issues/845

I asked many questions to @Fi3 as DMs over Discord. But I thought I should transfer those Q&A over here so they can be also used as future reference to the community.

**Comment by plebhash** [comment]
Could you provide a 1 sentence description for the following features:
- `with_serde`
- `with_buffer_pool`

**Reply by plebhash** [reply]
answer by @Fi3 :

> with_serde use binary_sv2 with serde
with_buffer_pool use a more efficient method to allocate instead of a vec to dig into it you should go in utils/buffer
in buffer/utils readme is explained if you are curios

**Comment by plebhash** [comment]
I'm quite confused by [this](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L324):
```
/// A frame can be either
/// 1: Sv2Frame
/// 2: NoiseFrame
/// 3: HandashakeFrame
///
#[derive(Debug)]
pub enum EitherFrame<T, B> {
    HandShake(HandShakeFrame),
    Sv2(Sv2Frame<T, B>),
}
```

`HandShakeFrame` is just an alias for `NoiseFrame`, so why does this comment list 3 items?

**Reply by plebhash** [reply]
answer by @Fi3:

> this seems wrong to me. I think that before changing the noise spec we had 3 frames so the code have been updted but not the comment let me check
at first look I can't find why
I dont' have time now to dig into it

> but right now is not like that for sure
before adding comment look at how is used in codec_sv2 to be sure
If I reember correctly noise_frame != handshake_frame. but not 100% sure maybe naming should be changed 
cause in the noise handshake the first mnessages are not framed so maybe for them we are not using the frame crates
but double check it

**Reply by plebhash** [reply]
`HandshakeFrame` is an alias for `NoiseFrame`: https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L90

btw this line is still marked as `/// todo` on my WIP because I'm struggling to understand the purpose of this alias too... but that is probably related to my lack of knowledge about Noise protocol... I'm watching some YouTube videos about Noise today to improve that

**Reply by plebhash** [reply]
answer by @Fi3 :

> ok leave it as todo and then when I will be back in florence i can check if we have to remove it or no
can you send me where it is used? If I remeber correctly is to again for performance reason when you want to change the id but everything else is the same? (if that is the case is not the safest or best interface) Or maybe is there to make the compiler happy in some weird cases?

**Reply by plebhash** [reply]
there's some instances of usage of `HandshakeFrame` on `codec_sv2`:
- https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/codec-sv2/src/lib.rs#L49
- https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/codec-sv2/src/encoder.rs#L105
- https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/codec-sv2/src/decoder.rs#L155

and also `network_helpers`:
- https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/roles/roles-utils/network-helpers/src/lib.rs#L73

**Reply by Fi3** [reply]
Right now `Sv2Frame` is used both for noise and not noise context and `NoiseFrame`  aka `HandShakeFrame` is used only during the handshake. Before the changes to the noise part of the protocol we used `NoiseFrame` for the noise context and the handshake part and `Sv2Frame` for the not noise context.
Right now the only thing that `NoiseFrame` does is to return the inner bytes without changing it (it can just be a `Vec`)

Given that `EitherFrame` could be something like that:

```rust
#[derive(Debug)]
pub enum EitherFrame<T, B> {
    HandShake(Vec<u8>) // no need to use buffer_sv2::Slice here imo
    Sv2(Sv2Frame<T, B>),
}
```


The type `Sv2Frame` represent an unencrypted stratum v2 frame.

**Reply by plebhash** [reply]
OK thanks, I'll make sure the documentation for the current state of the crate provides some description of this.

But IMO even with docs, it will be quite confusing for the user, and `framing_sv2` should be eventually refactored to make it easier to use and understand.

This is a future discussion, but I'm taking notes for a future refactoring.

Along with the total removal of `Frame` trait, I would propose that we:
- remove `HandShakeFrame` alias
- rename `NoiseFrame` to `NoiseHandShakeFrame`
- refactor APIs of `NoiseHandShakeFrame` to remove the impositions from `Frame` trait, making it more intuitive

**Comment by plebhash** [comment]
I'm trying to understand the implementation of [`Sv2Frame::from_bytes` and `Sv2Frame::from_bytes_unchecked`](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L152-L176).

Do we assume the `bytes` parameter is already serialized?

**Reply by plebhash** [reply]
answer by @Fi3 :

> exactly bytes is already serialized you build the frame from bytes when you receive a raw header + payload from the network. This do not deserialize the bytes and just build a frame that have the correct bytes len given an header. We do lazy deserialization cause a valid use case for a proxy is to just relay a frame as it is and would be a waste of resource to deserialize and reserilaize it. 
of course header need to be deserilized and the message to be decrypted so this thing are always done
btw deserilizing header is very fast as it is just a memcopy

**Comment by plebhash** [comment]
I understand most of what's happening on [`Sv2Frame::serialize`](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L111-L112).

But since this function is called `serialize`, shouldn't it be populating `self.serialized` on the `else if let Some(payload) = self.payload` branch?

If not here, where is `self.serialized` initialized?

**Reply by plebhash** [reply]
answer by @Fi3 :

> so serialized is populated and used only when you get raw bytes from the network. When you instead build a frame in you application you build it creating an header and a payload, where the payload is:
`T: Serialize + GetSize`

> When you call serialize you are ready to send the serailized bytes so you don't the frame anymore there is no need to populate serailized in this case. What you do is to writye the serailized bytes in a buffer provioded by the caller, this allow some nice optimization as using buffer_pool.

**Reply by jbesraa** [reply]
@Fi3 Do you think it makes sense to change Sv2Frame to an enum? 
Current version
```rust
/// Abstraction for a SV2 Frame.
#[derive(Debug, Clone)]
pub struct Sv2Frame<T, B> {
    header: Header,
    payload: Option<T>,
    /// Serialized header + payload
    serialized: Option<B>,
}
```
Enum version:
```rust
/// Abstraction for a SV2 Frame.
#[derive(Debug, Clone)]
pub enum Sv2Frame<T, B> {
    Raw(serialized: B),
    Frame {
        header: Header,
        payload: T,
    },
}

```

**Reply by Fi3** [reply]
I guess that is not an enum for performance reason but I don't remember (could be for other reasons also reasons that now do not exist anymore). I don't see why we need to change it  since it work and is not public. If you want to try I can take a look at the PR, if you can put it in a separate PR or commit I will really appreciate it.

**Reply by jbesraa** [reply]
here you go https://github.com/jbesraa/stratum/pull/7/commits/29f5e44dd0628c6698a813709f7dfe46d286398e

I think its much more cleaner with `enum`. What performance advantages structs has over enums?

**Reply by Fi3** [reply]
it is just a guess in fact I say also *could be for other reasons also reasons that now do not exist anymore*

**Reply by Fi3** [reply]
and ofc is not enum vs struct is that in that particular case would have made sense to structure the code in that way. I want to be clear that I'm not saying that enum are in someway magically slower the struct.

**Comment by plebhash** [comment]
what is the purpose of [`fn build_noise_frame_header(`](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L226) ?

it is not used anywhere

**Reply by plebhash** [reply]
answer by @Fi3:

> if is not used anywhere is likley something that we used and the we stop using and no one notice we can remove it in that case

**Comment by plebhash** [comment]
what is the purpose of [`Sv2Frame::map`](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/protocols/v2/framing-sv2/src/framing2.rs#L18)? 


I see some use-cases inside `roles` (e.g.: [this](https://github.com/stratum-mining/stratum/blob/546ecbc4b471a00b04b4376cb84e805f45b6d411/roles/jd-client/src/lib/upstream_sv2/upstream.rs#L361)) but it seems weird that you would only modify `payload` whithout modifying `serialized` or `header`. If we modify the payload but `serialized` and `header` stay the same, wouldn't that break the frame? I mean, now `payload` represents some new information.

**Reply by plebhash** [reply]
answer by @Fi3 :

> you never have both serialized and payload or one or the other. If you don't modify the len of the payload and the message type is the same you don't have to modify the header
btw yes is you have both serialized and payload you can not change just one
maybe we should at least check for it in the map and call it try_map? but before lets' understand the real reason fro why is there maybe we can remove it or change it with somthing else

**Comment by plebhash** [comment]
do you know the history behind this `unreachable!` macro?

https://github.com/stratum-mining/stratum/blob/24b8b1d8666d6055c4a964d2aac8ef2445734a10/protocols/v2/framing-sv2/src/framing2.rs#L92-L97

I commented out this code block and built the roles workspace without issues. Maybe it could be removed?

**Reply by plebhash** [reply]
answer by @Fi3 :

> if everything build with all feature (look at message generator) we can remove it

**Reply by plebhash** [reply]
I successfully executed `message-generator-tests.sh` after commenting out this code block.

I guess that's solid confirmation it can be removed

**Reply by Fi3** [reply]
yep is that is the case I guess that we can safely remove it

**Comment by plebhash** [comment]
the usage of generic `B` and the type alias `Slice` is a bit confusing... it seems both are aiming to achieve very similar goals but maybe in a redundant way

- `Sv2Frame<T, B>` has generic `B` to allow the user to decide which kind of buffer they want (which could be anything, as long as it satisfies the trait contraints).
-  `HandShakeFrame` uses the `Slice` alias to choose between `Vec<u8>` vs `buffer_sv2::Slice`, and this choice is enacted via the `with_buffer_pool`

wouldn't it be better if we simply replaced `B` with `Slice` everywhere?

generic `B` feels a bit overkill, since we already have two options available

@Fi3

**Reply by plebhash** [reply]
or alternatively, if keeping generic `B` is highly desirable for some reason, we could also make `HandShakeFrame` use generic `B` instead of type alias `Slice`

it just feels wrong to mix the two approaches for `Sv2Frame` and `HandShakeFrame`... since both are flavors of `enum Frame`, the API should allow the user to choose the buffer strategy with the same mechanism (either generic `B` or feature flag `with_buffer_pool`, but not both)

**Reply by Fi3** [reply]
rust best practice is to constrains generics only where needed. For example:
```rust
// You do not constrain G cause you don't need it to be constrained
struct Something<G> {
    first: G
}

// Here you need to add it so you do it and you constrain to Addable that is a trait that I just made up
impl<G: Addable> Something<G> {
    fn double(&mut self) {
        *self.first += self.first
    }
}
```

The example above do not make sense but I want to use it explain what I mean. 

Another best practice is to use explicative generics names. If `Sv2Frame<T, B>` is not clead you can use something like `Sv2Frame<Sv2Message, Buffer>` but to me it seems an overkill

**Reply by plebhash** [reply]
no but my main question is not really around generic constrains, that's secondary

the point I'm making is about using generic `B`/`Buffer` vs `with_buffer_pool`

it is confusing to have two different methods of choice in the same crate

I feel we should aim to have only one way that the user can choose which flavor of Buffer they prefer (either generic OR feature flag, ideally not both)

**Reply by Fi3** [reply]
in that case I would go with the feature flag only

**Reply by plebhash** [reply]
great, I also feel this is the best choice

cross referencing: https://github.com/stratum-mining/stratum/issues/903#issuecomment-2358882527
